File: backend/src/main/java/com/yourcompany/onlineexam/controller/ExamTimeController.java
package com.yourcompany.onlineexam.controller;

import com.yourcompany.onlineexam.service.ExamTimeService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/exam-time")
@CrossOrigin(origins = {"http://localhost:5173", "http://localhost:3000", "https://doanliennganh.vercel.app"})
public class ExamTimeController {
    private final ExamTimeService examTimeService;
    private static final Logger logger = LoggerFactory.getLogger(ExamTimeController.class);

    public ExamTimeController(ExamTimeService examTimeService) {
        this.examTimeService = examTimeService;
    }

    /**
     * Kiểm tra thời gian hiện tại và trạng thái bài thi
     */
    @GetMapping("/current")
    public ResponseEntity<Map<String, Object>> getCurrentTime() {
        try {
            Map<String, Object> timeInfo = examTimeService.getCurrentTimeInfo();
            return ResponseEntity.ok(timeInfo);
        } catch (Exception e) {
            logger.error("Lỗi khi lấy thông tin thời gian hiện tại: ", e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }

    /**
     * Kiểm tra trạng thái bài thi theo thời gian
     */
    @GetMapping("/{partId}/time-status")
    public ResponseEntity<Map<String, Object>> getTimeStatus(@PathVariable String partId) {
        try {
            Map<String, Object> status = examTimeService.getTimeStatus(partId);
            return ResponseEntity.ok(status);
        } catch (Exception e) {
            logger.error("Lỗi khi kiểm tra trạng thái thời gian bài thi {}: ", partId, e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }

    /**
     * Lấy thông tin đếm ngược thời gian
     */
    @GetMapping("/{partId}/countdown")
    public ResponseEntity<Map<String, Object>> getCountdown(@PathVariable String partId) {
        try {
            Map<String, Object> countdown = examTimeService.getCountdown(partId);
            return ResponseEntity.ok(countdown);
        } catch (Exception e) {
            logger.error("Lỗi khi lấy thông tin đếm ngược bài thi {}: ", partId, e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }

    /**
     * Kiểm tra xem có thể làm bài hay không
     */
    @GetMapping("/{partId}/can-take")
    public ResponseEntity<Map<String, Object>> canTakeExam(
            @PathVariable String partId,
            @RequestParam String userEmail) {
        try {
            Map<String, Object> result = examTimeService.canTakeExam(partId, userEmail);
            return ResponseEntity.ok(result);
        } catch (Exception e) {
            logger.error("Lỗi khi kiểm tra khả năng làm bài {}: ", partId, e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }

    /**
     * Lấy danh sách bài thi theo trạng thái thời gian
     */
    @GetMapping("/by-status")
    public ResponseEntity<Map<String, Object>> getExamsByTimeStatus(
            @RequestParam(required = false) String status,
            @RequestParam(required = false) String courseId) {
        try {
            Map<String, Object> result = examTimeService.getExamsByTimeStatus(status, courseId);
            return ResponseEntity.ok(result);
        } catch (Exception e) {
            logger.error("Lỗi khi lấy danh sách bài thi theo trạng thái thời gian: ", e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }
} 

File: backend/OnlineExam/Controllers/ExamTimeController.cs
using Microsoft.AspNetCore.Mvc;
using OnlineExam.Services;

namespace OnlineExam.Controllers;

[ApiController]
[Route("api/exam-time")]
public class ExamTimeController : ControllerBase
{
    private readonly ExamTimeService _service;
    private readonly ILogger<ExamTimeController> _logger;

    public ExamTimeController(ExamTimeService service, ILogger<ExamTimeController> logger)
    {
        _service = service;
        _logger = logger;
    }

    [HttpGet("current")]
    public ActionResult<Dictionary<string, object?>> GetCurrentTime()
    {
        try { return Ok(_service.GetCurrentTimeInfo()); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi lấy thời gian hiện tại"); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }

    [HttpGet("{partId}/time-status")]
    public async Task<ActionResult<Dictionary<string, object?>>> GetTimeStatus(string partId)
    {
        try { return Ok(await _service.GetTimeStatusAsync(partId)); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi kiểm tra time status {PartId}", partId); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }

    [HttpGet("{partId}/countdown")]
    public async Task<ActionResult<Dictionary<string, object?>>> GetCountdown(string partId)
    {
        try { return Ok(await _service.GetCountdownAsync(partId)); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi lấy countdown {PartId}", partId); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }

    [HttpGet("{partId}/can-take")]
    public async Task<ActionResult<Dictionary<string, object?>>> CanTakeExam(string partId, [FromQuery] string userEmail)
    {
        try { return Ok(await _service.CanTakeExamAsync(partId, userEmail)); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi kiểm tra can-take {PartId}", partId); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }

    [HttpGet("by-status")]
    public async Task<ActionResult<Dictionary<string, object?>>> GetExamsByTimeStatus(
        [FromQuery] string? status, [FromQuery] string? courseId)
    {
        try { return Ok(await _service.GetExamsByTimeStatusAsync(status, courseId)); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi lấy exams by status"); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }
}
